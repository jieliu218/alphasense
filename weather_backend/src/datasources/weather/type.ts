export enum CurrentDataParameters {
  Temperature2m = 'temperature_2m',
  WindSpeed10m = 'wind_speed_10m',
  RelativeHumidity2m = 'relative_humidity_2m',
}

export enum HourlyDataParameters {
  Temperature2m = 'temperature_2m',
  RelativeHumidity2m = 'relative_humidity_2m',
  WindSpeed10m = 'wind_speed_10m',
}

export interface QueryParameters {
  latitude: number
  longitude: number
  current: CurrentDataParameters[]
  hourly: HourlyDataParameters[]
}
