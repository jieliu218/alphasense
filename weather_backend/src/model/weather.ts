export interface WeatherDataModel {
  latitude: number
  longitude: number
  generationtime_ms: number
  utc_offset_seconds: number
  timezone: string
  timezone_abbreviation: string
  elevation: number
  current_units: CurrentUnitsModel
  current: CurrentWeatherModel
  hourly_units: HourlyUnitsModel
  hourly: HourlyWeatherModel
}

export interface CurrentUnitsModel {
  time: string
  interval: string
  temperature_2m: string
  wind_speed_10m: string
  relative_humidity_2m: string
}

export interface CurrentWeatherModel {
  time: string
  interval: number
  temperature_2m: number
  wind_speed_10m: number
  relative_humidity_2m: number
}

export interface HourlyUnitsModel {
  time: string
  temperature_2m: string
  relative_humidity_2m: string
  wind_speed_10m: string
}

export interface HourlyWeatherModel {
  time: string[]
  temperature_2m: number[]
  relative_humidity_2m: number[]
  wind_speed_10m: number[]
}
