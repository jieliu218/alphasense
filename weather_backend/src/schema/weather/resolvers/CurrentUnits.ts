import type { CurrentUnitsResolvers } from './../../types.generated';
export const CurrentUnits: CurrentUnitsResolvers = {
  temperature2m: (parent) => parent.temperature_2m,
  relativeHumidity2m: (parent) => parent.relative_humidity_2m,
  windSpeed10m: (parent) => parent.wind_speed_10m
};
