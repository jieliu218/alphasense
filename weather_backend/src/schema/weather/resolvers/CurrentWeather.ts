import type { CurrentWeatherResolvers } from './../../types.generated';
export const CurrentWeather: CurrentWeatherResolvers = {
  relativeHumidity2m: (parent) => parent.relative_humidity_2m,
  temperature2m: (parent) => parent.temperature_2m,
  windSpeed10m: (parent) => parent.wind_speed_10m
};
