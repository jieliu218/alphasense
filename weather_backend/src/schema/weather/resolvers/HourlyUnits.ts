import type { HourlyUnitsResolvers } from './../../types.generated';
export const HourlyUnits: HourlyUnitsResolvers = {
  temperature2m: (parent) => parent.temperature_2m,
  relativeHumidity2m: (parent) => parent.relative_humidity_2m,
  windSpeed10m: (parent) => parent.wind_speed_10m
};
