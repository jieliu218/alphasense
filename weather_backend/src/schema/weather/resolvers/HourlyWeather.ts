import type { HourlyWeatherResolvers } from './../../types.generated';
export const HourlyWeather: HourlyWeatherResolvers = {
  temperature2m: (parent) => parent.temperature_2m,
  relativeHumidity2m: (parent) => parent.relative_humidity_2m,
  windSpeed10m: (parent) => parent.wind_speed_10m
};
