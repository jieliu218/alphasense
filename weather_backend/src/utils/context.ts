import { type ApolloServer } from '@apollo/server';
import { type IncomingMessage } from 'http';
import type { Logger } from 'winston';

import { WeatherAPI } from '../datasources';

export class ContextValue {
  public dataSources: {
    weatherAPI: WeatherAPI
  };

  public logger: Logger;

  constructor ({
    server,
    logger
  }: {
    req: IncomingMessage
    server: ApolloServer<ContextValue>
    logger: Logger
  }) {
    const { cache } = server;
    this.dataSources = {
      weatherAPI: new WeatherAPI({ cache, contextValue: this })
    };
    this.logger = logger;
  }
}
