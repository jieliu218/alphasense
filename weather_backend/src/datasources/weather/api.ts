/*
 * The WeatherAPI class is a RESTDataSource that fetches weather data from the OpenMeteo API
 */
import { RESTDataSource } from '@apollo/datasource-rest';
import qs from 'qs';
import { ApolloServerErrorCode } from '@apollo/server/errors';
import { type KeyValueCache } from '@apollo/utils.keyvaluecache';
import { GraphQLError } from 'graphql';
import { HourlyDataParameters, CurrentDataParameters, type QueryParameters } from './type';
import { type ContextValue } from '../../utils/context';
import { type WeatherDataModel } from '../../model/weather';

export class WeatherAPI extends RESTDataSource {
  override baseURL = process.env.OPEN_METEO_BASE_URL ?? '';
  private readonly contextValue: ContextValue;

  /*
   * Constructor for the WeatherAPI class
   * @param options - The options to pass to the constructor
   */
  constructor (options: { contextValue: ContextValue, cache: KeyValueCache }) {
    super(options);
    this.contextValue = options.contextValue;
  }

  /*
   * Fetch weather data from the OpenMeteo API
   * @param longitude - The longitude of the location
   * @param latitude - The latitude of the location
   */
  async getWeatherData (longitude: number, latitude: number): Promise<WeatherDataModel> {
    const queryParameters: QueryParameters = {
      latitude,
      longitude,
      current: [
        CurrentDataParameters.Temperature2m,
        CurrentDataParameters.WindSpeed10m,
        CurrentDataParameters.RelativeHumidity2m
      ],
      hourly: [
        HourlyDataParameters.Temperature2m,
        HourlyDataParameters.RelativeHumidity2m,
        HourlyDataParameters.WindSpeed10m
      ]
    };
    const queryString = qs.stringify(queryParameters, { arrayFormat: 'comma' });
    return await this.fetchWithRetry(`forecast?${queryString}`, 3, 1000);
  }

  /*
   * Fetch data from the OpenMeteo API with retries
   * @param path - The path to fetch data from
   * @param retries - The number of retries left
   * @param delay - The delay in milliseconds before the next retry
   */
  async fetchWithRetry (path: string, retries: number, delay: number): Promise<WeatherDataModel> {
    try {
      return await this.get(path);
    } catch (error) {
      if (retries === 0) {
        this.contextValue.logger.error('WeatherAPI Retry Error', { error });
        throw new GraphQLError('An error occurred after fetching weather data retried 3 times', {
          extensions: {
            code: ApolloServerErrorCode.INTERNAL_SERVER_ERROR
          }
        });
      }
      this.contextValue.logger.info(
        `WeatherAPI: Retrying after ${delay}ms. Retries left: ${retries - 1}`
      );
      await new Promise((resolve) => setTimeout(resolve, delay));
      return await this.fetchWithRetry(path, retries - 1, delay);
    }
  }
}
