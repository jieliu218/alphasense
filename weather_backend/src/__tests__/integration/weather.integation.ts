import { ApolloServer } from '@apollo/server';
import { type Logger } from 'winston';
import { gql } from 'graphql-tag';
import { typeDefs } from '../../schema/typeDefs.generated';
import { resolvers } from '../../schema/resolvers.generated';
import { type TypeSource, type IResolvers } from '@graphql-tools/utils';
import { type ContextValue } from '../../utils/context';
import { WeatherAPI } from '../../datasources';
import { weatherDummy } from '../../__dummy__/weather.dummy';

const weatherAPI = new WeatherAPI({ contextValue: {} as unknown as ContextValue, cache: {} as any });
const mockLogger = {
  error: jest.fn(),
  info: jest.fn()
} as unknown as Logger;

const GET_WEATHER_DATA = gql`
  query GetWeatherData($latitude: Float!, $longitude: Float!) {
    getWeatherData(latitude: $latitude, longitude: $longitude) {
      latitude
      longitude
      generationtimeMs
      utcOffsetSeconds
      timezone
      timezoneAbbreviation
      elevation
      currentUnits {
        time
        interval
        temperature2m
        windSpeed10m
        relativeHumidity2m
      }
      current {
        time
        interval
        temperature2m
        windSpeed10m
        relativeHumidity2m
      }
      hourlyUnits {
        time
        temperature2m
        relativeHumidity2m
        windSpeed10m
      }
      hourly {
        time
        temperature2m
        relativeHumidity2m
        windSpeed10m
      }
    }
  }
`;

// Initialize ApolloServer
const testServer = new ApolloServer<ContextValue>({
  typeDefs: typeDefs as TypeSource,
  resolvers: resolvers as IResolvers<any>
});

describe('WeatherData Integation Test', () => {
  it('returns weather data success', async () => {
    weatherAPI.getWeatherData = jest.fn().mockResolvedValue({ ...weatherDummy });

    const response = await testServer.executeOperation(
      {
        query: GET_WEATHER_DATA,
        variables: { latitude: 40, longitude: -74 }
      },
      {
        contextValue: { logger: mockLogger, dataSources: { weatherAPI } }
      }
    );
    expect(response.body).toMatchSnapshot();
  });

  it('throws an error for invalid latitude', async () => {
    weatherAPI.getWeatherData = jest.fn().mockResolvedValue({ ...weatherDummy });

    const response = await testServer.executeOperation(
      {
        query: GET_WEATHER_DATA,
        variables: { latitude: 91, longitude: -74 }
      },
      {
        contextValue: { logger: mockLogger, dataSources: { weatherAPI } }
      }
    );
    expect(response.body).toMatchSnapshot();
  });

  it('throws an error for invalid longitude', async () => {
    weatherAPI.getWeatherData = jest.fn().mockResolvedValue({ ...weatherDummy });

    const response = await testServer.executeOperation(
      {
        query: GET_WEATHER_DATA,
        variables: { latitude: 40, longitude: 181 }
      },
      {
        contextValue: { logger: mockLogger, dataSources: { weatherAPI } }
      }
    );
    expect(response.body).toMatchSnapshot();
  });

  it('handles errors from the weatherAPI.getWeatherData method', async () => {
    weatherAPI.getWeatherData = jest.fn().mockResolvedValue(new Error('Network error'));

    const response = await testServer.executeOperation(
      {
        query: GET_WEATHER_DATA,
        variables: { latitude: 40, longitude: -74 }
      },
      {
        contextValue: { logger: mockLogger, dataSources: { weatherAPI } }
      }
    );
    expect(response.body).toMatchSnapshot();
  });
});
