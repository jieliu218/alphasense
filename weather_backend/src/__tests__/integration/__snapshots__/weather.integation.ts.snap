// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherData Integation Test handles errors from the weatherAPI.getWeatherData method 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getWeatherData": null,
    },
    "errors": [
      {
        "extensions": {
          "code": "INTERNAL_SERVER_ERROR",
        },
        "locations": [
          {
            "column": 3,
            "line": 2,
          },
        ],
        "message": "Network error",
        "path": [
          "getWeatherData",
        ],
      },
    ],
  },
}
`;

exports[`WeatherData Integation Test returns weather data success 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getWeatherData": {
        "current": {
          "interval": 900,
          "relativeHumidity2m": 59,
          "temperature2m": 0.4,
          "time": "2024-03-03T09:00",
          "windSpeed10m": 18.5,
        },
        "currentUnits": {
          "interval": "seconds",
          "relativeHumidity2m": "%",
          "temperature2m": "°C",
          "time": "iso8601",
          "windSpeed10m": "km/h",
        },
        "elevation": 250,
        "generationtimeMs": 0.102996826171875,
        "hourly": {
          "relativeHumidity2m": [
            82,
            85,
            86,
          ],
          "temperature2m": [
            -2.4,
            -2.6,
            -2.6,
          ],
          "time": [
            "2024-03-03T00:00",
            "2024-03-03T01:00",
            "2024-03-03T02:00",
          ],
          "windSpeed10m": [
            8.6,
            9.9,
            12.8,
          ],
        },
        "hourlyUnits": {
          "relativeHumidity2m": "%",
          "temperature2m": "°C",
          "time": "iso8601",
          "windSpeed10m": "km/h",
        },
        "latitude": 60,
        "longitude": 60,
        "timezone": "GMT",
        "timezoneAbbreviation": "GMT",
        "utcOffsetSeconds": 0,
      },
    },
    "errors": undefined,
  },
}
`;

exports[`WeatherData Integation Test throws an error for invalid latitude 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getWeatherData": null,
    },
    "errors": [
      {
        "extensions": {
          "code": "BAD_USER_INPUT",
        },
        "locations": [
          {
            "column": 3,
            "line": 2,
          },
        ],
        "message": "Latitude must be between -90 and 90 degrees.",
        "path": [
          "getWeatherData",
        ],
      },
    ],
  },
}
`;

exports[`WeatherData Integation Test throws an error for invalid longitude 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "getWeatherData": null,
    },
    "errors": [
      {
        "extensions": {
          "code": "BAD_USER_INPUT",
        },
        "locations": [
          {
            "column": 3,
            "line": 2,
          },
        ],
        "message": "Longitude must be between -180 and 180 degrees.",
        "path": [
          "getWeatherData",
        ],
      },
    ],
  },
}
`;
